[GStreamer][Holepunch] Add platform dependant bits

calvaris opened the pull request:
#### f1958b25b05493c9de01811ecbbad32704e75ddf
<pre>
[GStreamer][Holepunch] Add platform dependant bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=252241">https://bugs.webkit.org/show_bug.cgi?id=252241</a>

Reviewed by Philippe Normand.

Patch by Miguel Gómez &lt;magomez@igalia.com&gt;.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin):
(WebCore::setRectangleToVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::createHolePunchVideoSink):
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/260439@main">https://commits.webkit.org/260439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d84351688fe2c83e3c60f253ceb42a6c78491bac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41164 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117416 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116801 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8670 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100516 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114068 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97353 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42070 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28997 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97492 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10230 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30341 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8343 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7245 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16379 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49938 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105900 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7221 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12555 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26210 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1797,14 +1797,38 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
         GstState newState;
         gst_message_parse_state_changed(message, &currentState, &newState, nullptr);
-
         GST_DEBUG_OBJECT(pipeline(), "Changed state from %s to %s", gst_element_state_get_name(currentState), gst_element_state_get_name(newState));
```
Move this to just before the dump code, as it was before, please.

calvaris replied:
I don't follow. This code does not change. Should I remove the extra line removal?

philn replied:
No, sorry. I was confused :)
One small issue with this change though is that we'd get "Changed state from..." logs for all elements of the pipeline now. Can you move this log to after the early `break`?

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1797,14 +1797,38 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
+        // Construct a filename for the graphviz dot file output.
+        CString dotFileName = makeString(GST_OBJECT_NAME(pipeline()), '.',
+            gst_element_state_get_name(currentState), '_', gst_element_state_get_name(newState)).utf8();
+        GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS(GST_BIN(pipeline()), GST_DEBUG_GRAPH_SHOW_ALL, dotFileName.data());
```
Actually, this is not needed, should be done already in the simple bus message handler in GStreamerCommon.

philn commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -110,6 +110,13 @@ WEBKIT_OPTION_DEPEND(ENABLE_DOCUMENTATION ENABLE_INTROSPECTION)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_INTEL_CE "Whether to enable support for the Intel CE" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_RPI "Whether to enable support for the raspberry pi" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WESTEROS_SINK "Westeros-Sink to be used as video-sink for GStreamer video player" PUBLIC OFF)
+
```
Isn't this upstreamed already?

calvaris replied:
No. We had added depends for them to `ENABLE_VIDEO` but they were not defined.

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2788,11 +2812,13 @@ void MediaPlayerPrivateGStreamer::createGSTPlayBin(const URL& url)
     if (!m_player->isVideoPlayer())
         return;
 
+#if !USE(GSTREAMER_HOLEPUNCH)
```
Why?

calvaris replied:
@magomez will be able to answer better than me.

magomez replied:
Probably added at some point because it was causing a problem on some platform, and then was kept for every release. Feel free to remove it.

magomez replied:
Ah, got it. At that point there might not be a videoSink (if the platform uses autovideosink). Also, when using holepunch we don't take into account the size and orientation of the video, as the rectangle area is given only by the size of the video element, not the video frames. So when using holepunch we don't care about the video caps change.

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -110,6 +110,13 @@ WEBKIT_OPTION_DEPEND(ENABLE_DOCUMENTATION ENABLE_INTROSPECTION)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_WESTEROS "Whether to enable support for the Westeros platform" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_BCM_NEXUS "Whether to enable support for the Broadcom Nexus platform" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_INTEL_CE "Whether to enable support for the Intel CE" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_RPI "Whether to enable support for the raspberry pi" PUBLIC OFF)
```
Nit: capitalization:

```suggestion
WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_RPI "Whether to enable support for the Raspberry Pi" PUBLIC OFF)
```

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1797,14 +1797,33 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
 
+#if USE(GSTREAMER_HOLEPUNCH)
+        if (currentState == GST_STATE_NULL && newState == GST_STATE_READY) {
+            // If we didn't create a video sink, store a reference to the created one.
+            if (!m_videoSink) {
+                // Detect the videoSink element. Getting the video-sink property of the pipeline requires
+                // locking some elements, which may lead to deadlocks during playback. Instead, identify
+                // the videoSink based on its metadata.
+                GstElement* element = GST_ELEMENT(GST_MESSAGE_SRC(message));
+                if (GST_IS_BASE_SINK(element)) {
+                    const gchar* klassStr = gst_element_get_metadata(element, "klass");
+                    if (strstr(klassStr, "Sink") && strstr(klassStr, "Video")) {
+                        m_videoSink = element;
+
+                        // Ensure that there's a buffer with the transparent rectangle available when playback is going to start.
+                        pushNextHolePunchBuffer();
+                    }
+                }
+            }
```
Isn't this... dead code? In which case is `m_videoSink` null, exactly? I see `createVideoSink()` unconditionally called from `createGSTPlayBin()`.

philn replied:
Ok I see now this is called only for `USE(WPEWEBKIT_PLATFORM_BCM_NEXUS)`. Can you add it to the ifdef?

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3780,9 +3811,21 @@ GstElement* MediaPlayerPrivateGStreamer::createHolePunchVideoSink()
+#if USE(WESTEROS_SINK)
+    // Westeros using holepunch.
+    GRefPtr<GstElementFactory> westerosfactory = adoptGRef(gst_element_factory_find("westerossink"));
+    GstElement* videoSink = gst_element_factory_create(westerosfactory.get(), "WesterosVideoSink");
```
This can be simplified to `makeGStreamerElement("westerossink", "WesterosVideoSink")`

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3780,9 +3811,21 @@ GstElement* MediaPlayerPrivateGStreamer::createHolePunchVideoSink()
+    return nullptr;
+#endif
+
+    return gst_element_factory_make("fakevideosink", nullptr);
```
Can we keep using `makeGStreamerElement()` please?

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3780,9 +3811,21 @@ GstElement* MediaPlayerPrivateGStreamer::createHolePunchVideoSink()
+
+#if USE(WPEWEBKIT_PLATFORM_BCM_NEXUS)
+    // Nexus boxes use autovideosink.
+    return nullptr;
```
Are we sure here that returning NULL will make playbin, thus playsink fallback to autovideosink?

philn replied:
Ah yes, seems to be the case in `gen_video_chain()`.

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -110,6 +110,12 @@ WEBKIT_OPTION_DEPEND(ENABLE_DOCUMENTATION ENABLE_INTROSPECTION)
 WEBKIT_OPTION_CONFLICT(USE_SOUP2 ENABLE_WPE_1_1_API)
 
+# Supported platforms.
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_WESTEROS "Whether to enable support for the Westeros platform" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_BCM_NEXUS "Whether to enable support for the Broadcom Nexus platform" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WPEWEBKIT_PLATFORM_RPI "Whether to enable support for the Raspberry Pi" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_WESTEROS_SINK "Westeros-Sink to be used as video-sink for GStreamer video player" PUBLIC OFF)
```
Are you sure you want these to be build options rather than runtime checks? It would seem pretty unfortunate for standard builds of WPE WebKit to not include the Raspberry Pi support, in particular, since those devices are very popular.

calvaris replied:
Yes.

calvaris committed:
[GStreamer][Holepunch] Add platform dependant bits
https://bugs.webkit.org/show_bug.cgi?id=252241

Reviewed by Philippe Normand.

Patch by Miguel Gómez <magomez@igalia.com>.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin):
(WebCore::setRectangleToVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::createHolePunchVideoSink):
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/260439@main
